File: files/300-399/382/CH-382.R
library(tidyverse)
library(readxl)

path <- "300-399/382/CH-382 Filter.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "F3:F9")

is_valid = function(s) {
  nums = str_extract_all(s, "[1-9]")[[1]] %>% as.numeric()
  num_prod = prod(nums)
  num_sum = sum(nums)
  return(num_prod %% 4 == 0 || num_sum %% 3 == 0)
}

result = input %>%
  filter(map_lgl(ID, is_valid))

all.equal(result$ID, test$ID)
# [1] TRUE

File: files/300-399/382/CH-382.py
import pandas as pd
import re
from math import prod

path = "300-399/382/CH-382 Filter.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=2, nrows=8)
test = pd.read_excel(path, usecols="F", skiprows=2, nrows=6)

def is_valid(s):
    nums = [int(d) for d in re.findall(r"[1-9]", str(s))]
    num_prod = prod(nums)
    num_sum = sum(nums)
    return num_prod % 4 == 0 or num_sum % 3 == 0

result = input[input["ID"].apply(is_valid)].reset_index(drop=True)

print(result["ID"].equals(test.iloc[:, 0]))
# True
